controllers/snow_machineconfig_controller.go

```diff
@@ -0,0 +1,110 @@
+		Client:        client,
+		Log:           log,
+		ClientBuilder: clientBuilder,
+		Validator:     validator,
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SnowMachineConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&anywherev1.SnowMachineConfig{}).
+		Complete(r)
+}
+
+// TODO: add here kubebuilder permissions as needed
+func (r *SnowMachineConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := r.Log.WithValues("snowMachineConfig", req.NamespacedName)
+
+	// Fetch the SnowMachineConfig object
+	snowMachineConfig := &anywherev1.SnowMachineConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, snowMachineConfig); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(snowMachineConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+		}
+		if err := patchHelper.Patch(ctx, snowMachineConfig, patchOpts...); err != nil {
+			log.Error(reterr, "Failed to patch snowmachineconfig")
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// There's no need to go any further if the SnowMachineConfig is marked for deletion.
+	if !snowMachineConfig.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, snowMachineConfig, log)
+	}
+
+	result, err := r.reconcile(ctx, snowMachineConfig, log)
+	if err != nil {
+		log.Error(err, "Failed to reconcile SnowMachineConfig")
+	}
+	return result, nil
+}
+
+func (r *SnowMachineConfigReconciler) reconcile(ctx context.Context, snowMachineConfig *anywherev1.SnowMachineConfig, log logr.Logger) (_ ctrl.Result, reterr error) {
+	// TODO: need to figure out how to load creds in controller
+	deviceClientMap, err := r.ClientBuilder.BuildSnowAwsClientMap(ctx)
+	if err != nil {
+		log.Error(err, "failed to build aws client map")
+		return ctrl.Result{}, err
+	}
+	// Setting the aws client map on every reconcile based on the secrets at that point of time
+	validator := r.Validator.WithAwsClients(deviceClientMap)
+	if err := validator.ValidateMachineDeviceIPs(ctx, snowMachineConfig); err != nil {
+		log.Error(err, "failed validating machine device ips")
+		return ctrl.Result{}, err
+	}
+	if err := validator.ValidateEC2ImageExistsOnDevice(ctx, snowMachineConfig); err != nil {
+		log.Error(err, "failed validating ec2 image on devices")
+		return ctrl.Result{}, err
+	}
+	if err := validator.ValidateEC2SshKeyNameExists(ctx, snowMachineConfig); err != nil {
+		log.Error(err, "failed validating ec2 ssh key name")
+		return ctrl.Result{}, err
+	}
+	// TODO: Add Defaulter for snow machine config to generate default ssh keys and store as secrets on cluster
```

Comment:
IMO we should skip this for now. We decided something similar for vsphere: we don't generate keys from the controller, the user needs to pass them.

If there are user's requests later, we can added.